docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,298 @@
+let clientContext = new cloud.Bucket() as "ctx";
+clientContext.addObject("ctx", Json.stringify(client.context))
+```
+
+If the runtime cost is acceptable, the context can even be directly resolved directly from the bucket during the client's runtime:
+
+```wing
+bring cloud;
+bring clients;
+
+let clientContext = new cloud.Bucket() as "ctx";
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    return Json.parse(clientContext.get("ctx"));
+  },
+);
+
+// hydrates the context during deployment
+clientContext.addObject("ctx", Json.stringify(client.context))
+```
+
+In cases where reading from the filesystem is difficult and the overhead or a runtime lookup is too high, the context can be passed in via an environment variable:
+
+```wing
+bring cloud;
+bring clients;
+bring util;
+
+let bucket = new cloud.Bucket() as "shared";
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    let ctxRaw = util.getEnv("WING_CLIENT_CONTEXT");
+    let ctxDecoded = util.fromBase64(ctxRaw);
+    return Json.parse(ctxDecoded);
+  },
+);
+
+client.add(lift { bucket: [put] });
+```
+
+Regardless of the approach used to resolve the context, using the generated client will work the same way:
+
+```ts
+import { Client } from "@acme/infra-client";
+
+// performs the context resolution
+const client = new Client();
+
+await client.bucket.put("key", "value");
+```
+
+The client code is the same as the generated javascript as if it were used within wing itself!
+
+## Implementation Details
+
+### Client context data
+
+The client context data is a JSON object that contains all the necessary information to connect to the resources.
+For now, the only information that needs to be stored are environment variables.
+This corresponds to the only API available on `IInflightHost`, which is `addEnvironment`, which is how these values are added:
+
+```json
+{
+  "version": "1",
+  "env": {
+    "BUCKET_HANDLE_a107cec5": "abcxyz"
+  }
+}
+```
+
+The client itself is effectively a proxy of an unseen/unavailable `IInflightHost`, so any expressivity that interface has must be handled by the client.
+
+### Permissions
+
+The client does not directly/automatically handle permissions needed for the arbitrary host it could run on.
+There is no expectation that the deployer of the host is also using wing so there is no shared system to transfer permission knowledge.
+
+For example, when using
+
+```wing
+client.lift(lift { bucket: [put] });
+```
+
+it would be helpful to retrieve the necessary information about the `bucket` and the `put` operation.
+
+If you're using terraform and deploying to AWS, you will need the bucket name and the correspond IAM policy.
+There is no abstract API available to store/retrieve this information currently as this is handled only by target-specific subtypes of `IInflightHost`.
```

Comment:
What's the benefit of this approach over the current one in this proposal?

The end result is the same either way (you can get the necessary policy data) but the DX of `aws.generatePolicyStatements(client.lifts)` is both easy and powerful to me. I think it helps convey "You need to take care of making sure these permissions are applied." because you're the one that has to extract them in the first place.

The client is also otherwise completely platform agnostic in the same way an inflight is (i.e. it's more of a set of instructions rather than a "resource"), which seems pretty useful.
I'm not over-indexing on this agnosticism, but I don't see the benefit of getting rid of it.

Reply:
I see two main differences between the proposed approach (let's call it "A") and my proposal (let's call it "B").

1. Approach A is not platform agnostic. From a layering perspective, the way I see it everything that has to do with IAM policies is a platform concern, which implies that we don't want it to be explicit in the application code because it is different across platforms. Currently, the way to customize platformic concerns in Wing is either by creating a custom platform provider or using `wing.toml`. If we require the user to call `aws.generatePolicyStatements` explicitly in their Wing program, the program becomes platform-specific.

2. In approach A the user can just forget to create an IAM policy/role. It's on them to know somehow to call this method, generate the IAM policies and provision the resources. Approach B leads the user into the "pit of success".

I sense that there will be other aspects of these clients that are platform-specific, so it makes sense to design them like any other `cloud.*` resource and enable platform providers and extensions to control their behavior.

Reply:
> Approach A is not platform agnostic.
> If we require the user to call aws.generatePolicyStatements explicitly in their Wing program, the program becomes platform-specific.

Neither `A` nor `B` will result in a platform agnostic app. Even if a `B` creates a policy, some other AWS specific resource (i.e. role or user) will need to be created to attach the policy to. In many cases I expect this to be done in the same wing app.

> In approach A the user can just forget to create an IAM policy/role. It's on them to know somehow to call this method, generate the IAM policies and provision the resources. Approach B leads the user into the "pit of success".

In `B`, a policy is created but it's up to the user to use it. A policy does nothing on its own. Even worse, the policy is hidden so now creating a role to attach it to is a disconnected process.

Reply:
In approach B, the app code stays the same across different platforms. It will behave differently based on the target platform, and the behavior can be customized by the platform provider without needing to change the app code.

That's a fundamental design tenet for Wing and we should make sure we don't break the layering.

Reply:
In approach A, the app code stays the same across different platforms **and** it doesn't behave differently across platforms, so the there's no need to customize the behavior across platforms. This does not break any layering because there's no layer necessary.

To move this forward I will go ahead with your approach.

Reply:
OK, thanks! I think it will really help to flush out a couple of end-to-end examples and see how things look & feel.